File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/search_in_rotated_sorted_array/SearchinRotatedSortedArray2.java
package main.java.com.jiang.leetcode100.search_in_rotated_sorted_array;

import java.util.Scanner;

class Solution2 {
    public int search(int[] nums, int target) {

        int left = 0, right = nums.length - 1;
        while(left <= right) {
            int mid = (left + right) >> 1;
            System.out.println("mid: " + mid);
            if(nums[mid] == target) return mid;
            if(nums[mid] >= nums[0]) {
                if(nums[mid] > target && target >= nums[0]) right = mid - 1;
                else left = mid + 1;
            } else {
                if(nums[mid] < target && target <= nums[nums.length - 1]) left = mid + 1;
                else right = mid - 1;
            }
        }

        return -1;
    }
}

public class SearchinRotatedSortedArray2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        int target = scanner.nextInt();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }
        System.out.println(new Solution2().search(nums, target));
    }
}

File: Leetcode100/leetcode100_python/SearchinRotatedSortedArray2.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if not nums:
            return -1
        n = len(nums)
        left, right = 0, n - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] > nums[-1]:
                left = mid + 1
            else:
                right = mid - 1

        rot = left
        l, r = 0, n - 1
        while l <= r:
            m = (l + r) // 2
            realm = (m + rot) % n
            if nums[realm] == target:
                return realm
            if nums[realm] < target:
                l = m + 1
            else:
                r = m - 1

        return -1

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    target = int(input())
    solution = Solution()
    print(solution.search(nums, target))